Hide wiki markup as part of markdown-toggle-markup-hiding

jayemar opened the pull request:
When markup hiding is enabled, the markup for markdown-style links is hidden, but the markup for wiki links is not.  This code change fontifies the different sections of wiki links so that they are properly supported by `markdown-toggle-markup-hiding`.

## Related Issues

The changes address issues #557, #782, and #847.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] I have read the **CONTRIBUTING.md** document.
- [x] I have updated the documentation in the **README.md** file if necessary.
- [X] I have added an entry to **CHANGES.md**.
- [x] I have added tests to cover my changes.
- [X] All new and existing tests passed (using `make test`).

syohex commented on markdown-mode.el:
```diff
@@ -8174,6 +8229,33 @@ Translate filenames using `markdown-filename-translate-function'."
           (markdown--browse-url (markdown-link-url)))
     (user-error "Point is not at a Markdown link or URL")))
 
+(defconst markup-props (list 'invisible 'markdown-markup
```
Please add package prefix to global variables according to Emacs lisp naming convention

jayemar replied:
I fixed this and ran a linter to help catch some other issues

syohex commented:
- Please fix all byte-compile warnings
- Please add unit tests

jayemar commented:
I fixed the byte compiler warnings and updated the wiki link tests to check that the face properties are properly being set to support markup hiding. I think I've done everything that's called out in `CONTRIBUTING.md`, but if not let me know.

Wiki links seemed to be handled very different front inline and reference links.  I couldn't find a reason for the two different approaches, and it make some of the implementation more difficult than it needed to be, so some of my changes have the effect of treating wiki links much more like reference and inline links.  If I misunderstood this and there's a reason for the two different approaches please let me know.  (That's only if `markdown-enable-wiki-links` is `t`; otherwise they'll still ignored.)

jayemar commented:
I see this workflow is halted with the message that it is awaiting approval from a maintainer.  Is there anything I can do or need to do in order to get this branch through the build pipeline?

jayemar commented:
@syohex fixed merge conflicts and failing tests from recent updates to mouse face in links

syohex commented on markdown-mode.el:
```diff
@@ -3394,23 +3395,6 @@ the buffer)."
 (defun markdown-match-yaml-metadata-key (last)
   (markdown-match-propertized-text 'markdown-metadata-key last))
 
-(defun markdown-match-wiki-link (last)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8451,39 +8510,6 @@ See `markdown-wiki-link-p' and `markdown-follow-wiki-link'."
       (markdown-follow-wiki-link (markdown-wiki-link-link) arg)
     (user-error "Point is not at a Wiki Link")))
 
-(defun markdown-highlight-wiki-link (from to face)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8451,39 +8510,6 @@ See `markdown-wiki-link-p' and `markdown-follow-wiki-link'."
-  "Highlight the wiki link in the region between FROM and TO using FACE."
-  (put-text-property from to 'font-lock-face face))
-
-(defun markdown-unfontify-region-wiki-links (from to)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8451,39 +8510,6 @@ See `markdown-wiki-link-p' and `markdown-follow-wiki-link'."
-    ;; undo that if it wasn't originally marked modified
-    (set-buffer-modified-p modified)))
-
-(defun markdown-fontify-region-wiki-links (from to)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8502,50 +8528,6 @@ newline after."
         (setq new-to (point)))
     (cl-values new-from new-to)))
 
-(defun markdown-check-change-for-wiki-link (from to)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8502,50 +8528,6 @@ newline after."
-           (buffer-modified-p)
-           (set-buffer-modified-p nil)))))
-
-(defun markdown-check-change-for-wiki-link-after-change (from to _)
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8502,50 +8528,6 @@ newline after."
-Designed to be used with the `after-change-functions' hook."
-  (markdown-check-change-for-wiki-link from to))
-
-(defun markdown-fontify-buffer-wiki-links ()
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex commented on markdown-mode.el:
```diff
@@ -8556,33 +8538,10 @@ and disable it otherwise."
+             (if markdown-enable-wiki-links "enabled" "disabled")))
   (markdown-reload-extensions))
 
-(defun markdown-setup-wiki-link-hooks ()
```
Please don't remove public interfaces, this causes breaking backward compatibility

syohex reviewed (commented):
Please don't remove public interfaces  all of sudden.  When we remove public interfaces, fist we have to set them as obsoleted function/variable. Then after several version releases, then we will be able to remove them

jayemar commented:
@syohex Thank you for the feedback.  You're right that I should not have removed public functions so I've added them back in.

However, these functions were originally called via a hook that is no longer necessary as the markup for wiki links is now handled via `markdown-mode-font-lock-keywords` so as to function like other link types.  Given this, should I mark these functions as obsolete?

syohex commented:
> should I mark these functions as obsolete?

If it is no longer necessary or we can use another one instead of them, then please mark them as obsoleted variables/functions.

jayemar commented on markdown-mode.el:
```diff
@@ -278,6 +278,16 @@ project: search for wiki link targets under project root"
 (make-obsolete-variable 'markdown-wiki-link-search-subdirectories 'markdown-wiki-link-search-type "2.5")
 (make-obsolete-variable 'markdown-wiki-link-search-parent-directories 'markdown-wiki-link-search-type "2.5")
 
```
These functions are no longer used due to the way wiki links are handled with the changes

jayemar commented:
@syohex I have marked the functions that are no longer used/necessary as obsolete

syohex commented on markdown-mode.el:
```diff
@@ -278,6 +278,16 @@ This is the default search behavior of Ikiwiki."
 (make-obsolete-variable 'markdown-wiki-link-search-subdirectories 'markdown-wiki-link-search-type "2.5")
 (make-obsolete-variable 'markdown-wiki-link-search-parent-directories 'markdown-wiki-link-search-type "2.5")
 
+(make-obsolete 'markdown-match-wiki-link nil "2.3")
```
The 3rd argument of the `make-obsolete` means that **when** this function was made obsolete. So, it should be next release version(`2.8`), not past released versions

syohex commented:
Please update CHANGES.md

jayemar commented:
I've updated CHANGES.md to reference the changes from this branch.

And thanks for the changes to fix the byte compiler warnings!
